### PR TITLE
1255: Align order of displayed product block details with order of pr…

### PR DESCRIPTION
**Pull request by DutchBen**
…oduct block tree

### CODE REVIEW

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.spec.ts [hunk 1]
```diff
@@ -119,3 +120,239 @@ describe('getWfoTreeNodeDepth', () => {
...
+                "children": [],
+                "isOutsideCurrentSubscription": false
+            }
+        ],
+        "isOutsideCurrentSubscription": false
+    }`;
+
+    const tree = JSON.parse(json) as TreeBlock;
+
+    it('should return 1 for the root node', () => {
```
Comment (mrijk) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.spec.ts:304]: Looks a bit verbose

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts [hunk 2]
```diff
@@ -21,9 +21,18 @@ export function getWfoTreeNodeDepth(
...
 
 export const sortTreeBlockByLabel = (
-    { label: labelA }: TreeBlock,
-    { label: labelB }: TreeBlock,
+    { label: labelA, id: idA }: TreeBlock,
+    { label: labelB, id: idB }: TreeBlock,
 ): number => {
+    if (labelA === labelB) {
+        if (idA < idB) {
+            return -1;
```
Comment (mrijk) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:29]: I assume that only < and > are defined and you cant just do a 'return idA - idB' ?

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts [hunk 3]
```diff
@@ -32,3 +41,38 @@ export const sortTreeBlockByLabel = (
...
+        count += 1 + countChildeNodes(child);
+    }
+    return count;
+};
+
+export const getPositionInTree = (
+    tree: TreeBlock,
+    id: number,
+): number | undefined => {
+    let startPosition = 1;
```
Comment (mrijk) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:57]: A recursive solution would earn you bonus points and would get rid of the 'let' ;)

File: packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx [hunk 4]
```diff
@@ -97,6 +101,26 @@ export const WfoSubscriptionDetailTree = ({
...
+     * alphabetically per level we can not depend on the order of product block ids anymore so we sort by the position in the tree.
+     */
+    const sortByTree = (idA: number, idB: number) => {
+        if (!tree) return 0;
+
+        const positionA = getPositionInTree(tree, idA);
+        const positionB = getPositionInTree(tree, idB);
+        if (!positionA || !positionB) return 0;
+
+        if (positionA < positionB) {
```
Comment (mrijk) [inline packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx:115]: Subtraction is not defined on the positionA type? (so you could do return positionA - positionB)

Comment (ricardovdheijden) [reply packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx:115]: Yes this could be a subtraction indeed

File: packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx [hunk 5]
```diff
@@ -154,9 +178,8 @@ export const WfoSubscriptionDetailTree = ({
                         </EuiCallOut>
                     )}
                     {selectedIds.length !== 0 &&
-                        selectedIds.map((id, index) => {
-                            const block = idToNodeMap[selectedIds[index]];
-
+                        selectedIds.sort(sortByTree).map((id, index) => {
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx:181]: Replace `.sort()` with `.toSorted()` to avoid add behavior

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts [hunk 6]
```diff
@@ -32,3 +41,38 @@ export const sortTreeBlockByLabel = (
     }
     return 0;
 };
+
+const countChildeNodes = (node: TreeBlock): number => {
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:45]: `getChildrenNodesCount` or `getNumberOfChildrenNodes`

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts [hunk 7]
```diff
@@ -32,3 +41,38 @@ export const sortTreeBlockByLabel = (
...
+    let count = 0;
+    for (const child of node.children) {
+        count += 1 + countChildeNodes(child);
+    }
+    return count;
+};
+
+export const getPositionInTree = (
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:53]: Since this function is a bit complex, maybe add some explanation above

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts [hunk 8]
```diff
@@ -32,3 +41,38 @@ export const sortTreeBlockByLabel = (
...
+    }
+
+    for (const child of tree.children) {
+        startPosition += 1;
+        if (child.id === id) {
+            return startPosition;
+        }
+        if (child.children.length > 0) {
+            const childPosition = getPositionInTree(child, id);
+            if (childPosition) {
```
Comment (ricardovdheijden) [inline packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:69]: Consider to rewrite this, since it gets a bit hard to follow

**Review by ricardovdheijden — APPROVED** [review]

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.spec.ts (hunk 1 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.spec.ts:304]: The test data is an actual data set that showed me wrong behaviour. I wanted to make sure it passed completely. I think it doesn't hurt to have it here since its there already

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts (hunk 2 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:29]: I'm not sure what you mean by only < and > are defined? The sort function only needs to determine if numbers are  greater than, smaller than or equal to each other. Returning 1, -1, or 0 is just a convention I think is clearer

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts (hunk 3 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:57]: Damn.. missed points

File: packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx (hunk 4 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx:115]: The same answer as above ... I think

File: packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx (hunk 5 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoSubscription/WfoSubscriptionDetailTree.tsx:181]: In this case that is actual overkill. selectedIds is just an array of numbers.

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts (hunk 6 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:45]: changed it to the little more accurate countDescendantNodes. Count is still a verb so I think it works

File: packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts (hunk 7 above)
Comment (DutchBen) [reply packages/orchestrator-ui-components/src/components/WfoTree/treeUtils.ts:53]: Done
